// angular-auth-oidc-client/esm2022/lib/auto-login/auto-login.service.mjs
import { Injectable, inject } from '@angular/core';
import { Router } from '@angular/router';
import { StoragePersistenceService } from '../storage/storage-persistence.service';
import * as i0 from "@angular/core";
const STORAGE_KEY = 'redirect';
export class AutoLoginService {
    constructor() {
        this.storageService = inject(StoragePersistenceService);
        this.router = inject(Router);
    }
    checkSavedRedirectRouteAndNavigate(config) {
        if (!config) {
            return;
        }
        const savedRouteForRedirect = this.getStoredRedirectRoute(config);
        if (savedRouteForRedirect != null) {
            this.deleteStoredRedirectRoute(config);
            this.router.navigateByUrl(savedRouteForRedirect);
        }
    }
    /**
     * Saves the redirect URL to storage.
     *
     * @param config The OpenId configuration.
     * @param url The redirect URL to save.
     */
    saveRedirectRoute(config, url) {
        if (!config) {
            return;
        }
        this.storageService.write(STORAGE_KEY, url, config);
    }
    /**
     * Gets the stored redirect URL from storage.
     */
    getStoredRedirectRoute(config) {
        return this.storageService.read(STORAGE_KEY, config);
    }
    /**
     * Removes the redirect URL from storage.
     */
    deleteStoredRedirectRoute(config) {
        this.storageService.remove(STORAGE_KEY, config);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: AutoLoginService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: AutoLoginService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: AutoLoginService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
