// angular-auth-oidc-client/esm2022/lib/interceptor/closest-matching-route.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class ClosestMatchingRouteService {
    getConfigIdForClosestMatchingRoute(route, configurations) {
        for (const config of configurations) {
            const { secureRoutes } = config;
            for (const configuredRoute of secureRoutes ?? []) {
                if (route.startsWith(configuredRoute)) {
                    return {
                        matchingRoute: configuredRoute,
                        matchingConfig: config,
                    };
                }
            }
        }
        return {
            matchingRoute: null,
            matchingConfig: null,
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: ClosestMatchingRouteService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: ClosestMatchingRouteService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: ClosestMatchingRouteService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
